### PR TITLE
Improve rounding

### LINKED ISSUE #713: Popper is not correctly placed by `1px` for some viewport widths in `flexbox` container
### CodePen demo

https://codepen.io/anon/pen/WYOzLr

### Steps to reproduce the problem

1.  resize the codepen window

screen shot 2018-11-16 at 10 20 47

screen shot 2018-11-16 at 10 21 08

### What is the expected behavior?

Popper should be placed pixel perfect to reference element 

### What went wrong?

For some viewport widths we got in `computeStyle` `popper.left` very close to integer value, but `Math.flor` in this case will break the offset by 1px

```
popper.left 561.984375
offsets.left 561
```

https://github.com/FezVrasta/popper.js/blob/master/packages/popper/src/modifiers/computeStyle.js#L44 changing it back to `Math.round` looks like fixes the problem

### Any other comments?

Connected with https://github.com/FezVrasta/popper.js/issues/277

### PR DESCRIPTION
**Pull request by atomiks**
Fixes #713 

It appears that `Math.round()` is always preferred except with placements `top` and `bottom` (not a variation) and where `Math.floor()` is preferred (and only on the `left` and `right` values)...

See the playground: https://codepen.io/anon/pen/Pxjdoo

### CODE REVIEW

**Comment by FezVrasta** [comment]
Should the playground include already your fix?

I see some misalignment:

image

**Comment by atomiks** [comment]
Yeah, line 1243...

Perhaps we need to take into account the odd/even width then.

When I resize the window using PR code:

- pop = even, ref = odd === perfect
- pop = even, ref = even === jittery / sometimes incorrect position

When I resize the window always using Math.round():

- pop = even, ref = odd === jittery / sometimes incorrect position
- pop = even, ref = even === perfect

Can you reproduce that?

**Comment by FezVrasta** [comment]
Yes I can repro

**Comment by atomiks** [comment]
I think this is the pattern:

- If the reference & popper are the _same_ oddness in width, then use Math.round()
- If they're different oddness in width, then use Math.floor()

This still only applies to horizontal placement and the left/right values

File: packages/popper/src/modifiers/computeStyle.js [hunk 1]
```diff
@@ -38,13 +38,23 @@ export default function computeStyle(data, options) {
...
+  // For pixel-perfect positioning in horizontal placements, top/bottom prefers
+  // rounded values, while left/right prefers floored values.
+  // Always use `round` for variations (-start and -end).
+  const referenceAndPopperHaveSameWidthOddness = 
+    (reference.width % 2 === 0) === (popper.width % 2 === 0);
+  const toInteger =
+    ['left', 'right'].indexOf(data.placement) !== -1 ||
+    data.placement.indexOf('-') > -1 || 
+    referenceAndPopperHaveSameWidthOddness
+      ? Math.round
```
[the remark this answers is not in the document]
Comment (atomiks) [reply packages/popper/src/modifiers/computeStyle.js]: `ceil` causes some jitters from what I can see.

[the remark this answers is not in the document]
Comment (FezVrasta) [reply packages/popper/src/modifiers/computeStyle.js]: I see, thanks

**Comment by FezVrasta** [comment]
`top` and `bottom` are working fine as they are already?

**Comment by atomiks** [comment]
It appears so yes. Horizontally it appears flush with the reference, and you can change the code at the very bottom to `left / left-end` as an example too.

However I just noticed that `pop = odd, ref = even` doesn't like Math.round() (but doesn't jitter). Another edge case? 😓

**Comment by FezVrasta** [comment]
Please, let's make sure the behavior is consistent across browsers also.

**Comment by atomiks** [comment]
This might be getting stranger by the second..

If both are odd width, then we need

```js
left: toInteger(bothOddWidth ? popper.left - 1 : popper.left),
```

I'm going to check Firefox and Safari now

**Comment by FezVrasta** [comment]
Considering the number of edge cases you are finding I think it would make sense to write some tests so make sure everything is working as expected.

**Comment by atomiks** [comment]
On Firefox it looks like the element is positioned on a half pixel, no matter what if we use a full pixel then it can't be positioned correctly. My suggestion there would be _not to round_ if the screen is HiDPI (so there won't be blurriness). If you don't round at all, FF appears perfect, same with Safari.

**Comment by FezVrasta** [comment]
this thing is getting complex 😅

may we limit this PR to the most basic use case (non HiDPI screens) and think about the HiDPI screens in a new one?

**Comment by atomiks** [comment]
Yeah it's kinda crazy for something that should be simple..

FWIW here is what it would look like to check for that:

```js
const isVertical = ['left', 'right'].indexOf(data.placement) !== -1;
const isVariation = data.placement.indexOf('-') !== -1;
const sameWidthOddness = reference.width % 2 === popper.width % 2;
const bothOddWidth = reference.width % 2 === 1 && popper.width % 2 === 1;
const isHiDPIScreen = window.devicePixelRatio >= 2;
const noRound = v => v;

// We don't need to round if the device has a high pixel density
const horizontalToInteger = isHiDPIScreen
  ? noRound
  : isVertical || isVariation || sameWidthOddness
  ? Math.round
  : Math.floor;
const verticalToInteger = isHiDPIScreen ? noRound : Math.round;

const offsets = {
  left: horizontalToInteger(
    bothOddWidth && !isVariation && !isHiDPIScreen
      ? popper.left - 1
      : popper.left
  ),
  top: verticalToInteger(popper.top),
  bottom: verticalToInteger(popper.bottom),
  right: horizontalToInteger(popper.right)
};
```

And yes I agree it needs tests due to all the logic 😅

**Comment by atomiks** [comment]
Will try to write some tests in a sec

**Comment by atomiks** [comment]
@FezVrasta are the tests appropriate?
